rskj-core/src/main/java/co/rsk/peg/federation/FederationSupportImpl.java

```diff
@@ -588,83 +602,148 @@ private Integer addFederatorPublicKeyMultikey(boolean dryRun, BtcECKey btcKey, E
-        Instant creationTime = Instant.ofEpochMilli(rskExecutionBlock.getTimestamp());
         Federation activeFederation = getActiveFederation();
-        provider.setOldFederation(activeFederation);
-        provider.setNewFederation(
-            currentPendingFederation.buildFederation(
-                creationTime,
-                rskExecutionBlock.getNumber(),
-                constants,
-                activations
-            )
-        );
-        provider.setPendingFederation(null);
+        preserveFederationChangeInfo(activeFederation);
+        logCommitmentWithVotedFederation(eventLogger, activeFederation, proposedFederation);
 
-        // Clear votes on election
+        return FederationChangeResponseCode.SUCCESSFUL;
+    }
+
+    private Federation buildFederationFromPendingFederation(PendingFederation pendingFederation) {
+        Instant federationCreationTime = Instant.ofEpochMilli(rskExecutionBlock.getTimestamp());
+        long federationCreationBlockNumber = rskExecutionBlock.getNumber();
+
+        return pendingFederation.buildFederation(federationCreationTime, federationCreationBlockNumber, constants, activations);
+    }
+
+    private void clearPendingFederationVoting() {
+        // Clear pending federation and votes on election
+        provider.setPendingFederation(null);
         provider.getFederationElection(constants.getFederationChangeAuthorizer()).clear();
+    }
 
-        Federation oldFederation = provider.getOldFederation(constants, activations);
+    private void preserveFederationChangeInfo(Federation activeFederation) {
+        provider.setNextFederationCreationBlockHeight(rskExecutionBlock.getNumber());
 
-        if (activations.isActive(RSKIP186)) {
-            // Preserve federation change info
-            long nextFederationCreationBlockHeight = rskExecutionBlock.getNumber();
-            provider.setNextFederationCreationBlockHeight(nextFederationCreationBlockHeight);
-            Script oldFederationP2SHScript = activations.isActive(RSKIP377) && activeFederation instanceof ErpFederation ?
-                ((ErpFederation) activeFederation).getDefaultP2SHScript() : activeFederation.getP2SHScript();
-            provider.setLastRetiredFederationP2SHScript(oldFederationP2SHScript);
-        }
+        Script activeFederationMembersP2SHScript = getFederationMembersP2SHScript(activeFederation);
+        provider.setLastRetiredFederationP2SHScript(activeFederationMembersP2SHScript);
+    }
 
-        Federation newFederation = provider.getNewFederation(constants, activations);
+    protected Script getFederationMembersP2SHScript(Federation federation) {
+        // when the federation is a standard multisig,
+        // the members p2sh script is the p2sh script
+        if (!activations.isActive(RSKIP377)) {
+            return federation.getP2SHScript();
+        }
+        if (!(federation instanceof ErpFederation)) {
+            return federation.getP2SHScript();
+        }
```

Comment:
```suggestion
        if (!activations.isActive(RSKIP377) || !(federation instanceof ErpFederation)) {
            return federation.getP2SHScript();
        }
```
Why not just join?

Reply:
it just feels easier for me to understand the flow. But i can change it